### PR TITLE
Demand revoke_and_ack before more updates

### LINKED ISSUE #526: continuously send updates and `commitment_signed`s without receiving `revoke_and_ack`s.
Is it permissible to continuously send updates and `commitment_signed`s without receiving `revoke_and_ack`s?
There will be several valid unrevoked remote commitment transactions after that.
`revoke_and_ack`s will come late.
It seems that it is not clearly stated that prohibiting this to the sender.

The following sequence.
```
A --- update ---> B
A --- commit ---> B
A --- update ---> B
A --- commit ---> B
A <--- revoke --- B
A <--- revoke --- B
```

### PR DESCRIPTION
**Pull request by OrfeasLitos**
Fixes #526

### CODE REVIEW

File: 02-peer-protocol.md [hunk 1]
```diff
@@ -953,6 +953,9 @@ fee changes).
   to BIP69 lexicographic ordering of the commitment transaction.
   - if it has not recently received a message from the remote node:
       - SHOULD use `ping` and await the reply `pong` before sending `commitment_signed`.
+  - after sending a `commitment_signed` message:
+      - MUST NOT process any `update` message from the other party before receiving the
+        `revoke_and_ack` that corresponds to the `commitment_signed` sent.
```
Comment (cdecker) [inline 02-peer-protocol.md]: This should also clarify what happens if the counterparty continues to send updates. Should it queue them up? Should it fail the channel?

Comment (OrfeasLitos) [reply 02-peer-protocol.md]: I propose that further `update` messages should be simply ignored. I've updated the text accordingly.

**Comment by nayuta-gondo** [comment]
My explanation is not enough in #526.
I do not feel a problem with the `update_` messages before `revoke_and_ack` after `commitment_signed`.
I think they should be held until the next `commitment_signed`.
The problem is a new `commitment_signed` before `revoke_and_ack` returns.

**Comment by OrfeasLitos** [comment]
Failing the channel, queuing or ignoring received `update` messages are three options then. A fourth is to create a new message in the lines of `expecting_revocation`.

- Failing the channel is too harsh and would probably lead to channels being closed too often.
- I dislike queuing because it opens a DoS vector.
- Ignoring messages is the simplest solution given that there exist no edge cases where the two nodes reach a weird state because of bad timing of messages.
- If such an edge case does exist, then the new `expecting_revocation` message would save things.

*TL;DR*: I support ignoring messages, which is not too harsh and secure at the same time. If it leads to inconsistencies, then let's add a new `expecting_revocation` message.

**Comment by SomberNight** [comment]
If you ignore the message, how do you expect the sender of that message to realise it was ignored? I would not be surprised if ignoring the message would lead to some other protocol violation/inconsistency which then would result in the channel being closed/failed.

**Comment by OrfeasLitos** [comment]
Ignoring the message is a good idea in the following example: `A` sends `commitment_signed` but the message is delayed. `B` sends `update` before receiving `commitment_signed`. `B` then realizes that the latest `update` will be ignored and thus sends a `revoke_and_ack`.

However in this example, if `A`'s first message is dropped, then `B` will never receive `commitment_signed`. This will probably lead to the channel failing. So ignoring the `update` is probably a bad idea, you are right. A better solution for this example would be for `A` to resend `commitment_signed` to `B` upon receiving the `update`. I'm not sure if a simple retransmission of `commitment_signed` is enough, or the new message `expecting_revocation` is required to cover some other cases.

Skimming through https://github.com/lightningnetwork/lightning-rfc/blob/master/02-peer-protocol.md#message-retransmission , it is not clear whether the case we discuss is covered.

**Comment by TheBlueMatt** [comment]
I strongly think we should be failing the channel if a counterparty is misbehaving and sending multiple updates without us having sent an RAA. The problem is the "ACK" part of RAA - if they send multiple updates before our RAA then there is no way for us (or them) to have any idea what we're ACK'ing, and, thus, which updates should be included in our next commitment_signed. This is pretty obviously very busted behavior, so we should probably not be talking to a node like that.

**Comment by nayuta-gondo** [comment]
RAA's ACK will be ACK for CS?
It can be said that it is for the `update_`s committed by the CS.
`update_`s sent after a CA will not be committed by the CA.
In the time series, I think it is clear which `update_`s are included in which CS.

**Comment by TheBlueMatt** [comment]
To quote the intro section to revoke_and_ack:
```
This message also implicitly serves as an acknowledgment of receipt of the commitment_signed, so this is a logical time for the commitment_signed sender to apply (to its own commitment) any pending updates it sent before that commitment_signed.
```

**Comment by rustyrussell** [comment]
This is wrong.  You can certainly send more updated (and we do!) before receiving revoke_and_ack.  But you can't send another commitment_signed.   If you try to implement that you'll see why: you don't know the per-commitment point for the N+1'th commitment, so can't create the tx.

**Comment by OrfeasLitos** [comment]
But one question still remains: After sending `commitment_signed` and before receiving `revoke_and_ack`, what should a node do on _receiving_ an `update_`?
- Ignore it?
- Fail the channel?
- Drop it and resend `commitment_signed`?
- Queue it and wait for `revoke_and_ack`?
- A combination of the above?
It is important that this behavior be specified, unless we want nodes to implement it in incompatible ways.

**Comment by rustyrussell** [comment]
You apply it to your local commitment, as always.  It has nothing (much) to do with sending commitment_signed, which is about applying changes to the *remote* update.

**Comment by OrfeasLitos** [comment]
So, to sum it all up, a local node that has just sent a `commitment_signed` is waiting for a `revoke_and_ack` (which may come very far in the future, not a problem), and also accepts `update_`s from the remote and it applies them (adds `add` and removes `fulfill`) to its local commitment. It can also send new `update_`s (given that they don't fulfill non-irrevocably-commited htlcs). The one thing that the local node can't do is send another `commitment_signed`. Once it receives the expected `revoke_and_ack`, the local node may send a new `commitment_signed` with the signatures to all the htlcs that have been received and applied in the meanwhile.

Is this all correct?

**Comment by OrfeasLitos** [comment]
Closing because this PR is not constructive. I assume that the silence means that my previous comment was indeed correct.
